src/AsyncWriter.js

```diff
@@ -1,357 +1,184 @@
     }
-};
-
-function AsyncFragment(asyncWriter) {
-    Fragment.call(this, asyncWriter);
-}
-
-AsyncFragment.prototype = {
-    end: function () {
-        if (!this.finished) {
-            // Make sure end is only called once by the user
-            this.finished = true;
-
-            if (this.ready) {
-                // There are no nested asynchronous fragments that are
-                // remaining and we are ready to be flushed then let's do it!
-                this.flush();
-            }
-        }
-    },
-    flush: function () {
-        if (!this.finished) {
-            // Skipped Flushing since not finished
-            return;
-        }
-        this.flushed = true;
-        var writer = this.writer;
-        this.writer = this.asyncWriter.writer = voidWriter; // Prevent additional out-of-order writes
-        flushNext(this, writer);
-    }
-};
 
-function AsyncWriter(writer, global, async, events, buffer) {
     this.data = {};
-    this.global = this.attributes /* legacy */ = (global || (global = {}));
-    this._af = this._prevAF = this._parentAF = null;
-    this._isSync = false;
-    this._last = null;
-
-    if (!events) {
-        // Use the underlying stream as the event emitter if available.
-        // Otherwise, create a new event emitter
-        events = writer && writer.on ? writer : new EventEmitter();
-    }
-
-    this._events = global.events /* deprecated */ = events;
-
-    if (!async) {
-        async = {
-            remaining: 0,
-            ended: false,
-            last: 0,
-            finished: false
-        };
-    }
-
-    this._async = async;
-
-    var stream;
-
-    if (!writer) {
-        writer = new StringWriter(this._events);
-    } else if (buffer) {
-        stream = writer;
-        writer = new BufferedWriter(writer);
-    }
-
-    this.stream = stream || writer;
-    this.writer = this._stream = writer;
+    this.writer = writer;
+    writer.stream = this;
 }
 
-AsyncWriter.DEFAULT_TIMEOUT = 10000;
-
-AsyncWriter.prototype = {
-    constructor: AsyncWriter,
+AsyncStream.DEFAULT_TIMEOUT = 10000;
+AsyncStream.INCLUDE_STACK = typeof process !== 'undefined' && process.env.NODE_ENV === 'development';
+AsyncStream.enableAsyncStackTrace = function() {
+    AsyncStream.INCLUDE_STACK = true;
+};
 
-    isAsyncWriter: AsyncWriter,
+var proto = AsyncStream.prototype = {
+    constructor: AsyncStream,
+    isAsyncWriter: AsyncStream,
 
     sync: function() {
         this._isSync = true;
     },
-    getAttributes: function () {
-        return this.global;
-    },
-    getAttribute: function (name) {
-        return this.global[name];
-    },
+
     write: function (str) {
         if (str != null) {
             this.writer.write(str.toString());
         }
         return this;
     },
+
     getOutput: function () {
-        return this.writer.toString();
+        return this.writer === voidWriter ? this._originalWriter.toString() : this.writer.toString();
```

Comment:
@patrick-steele-idem I wanted to point this out and there might be one more change we want to make here.

In the previous implementation, the original `out` would eventually point back to the original writer, so `getOutput()` originally just returned `this.writer.toString()`.  However the current implementation sets **every** stream's writer to `voidWriter` once the stream ends.   

I'm thinking this should just be set to `return this._originalWriter.toString()`, because it feels a little weird to get the buffer contents and I'm thinking this is really only to be used once the stream has finished in which case this check is unnecessary.  

Here's an example of calling `getOutput()` before the stream finishes:

```js
var out = create();
out.w('1');
var asyncOut = out.beginAsync();
setTimeout(() => asyncOut.end('2'), 100);
out.w('3');
out.getContent(); // '3' in the current implementation, 
                  // '1' if we switch to `return this._originalWriter.toString()`
```